4.x: Exception when writing in HTTP/2

Helidon 4.4.1
Linux
benchmark

Sometimes when writing byte array to a server response in HTTP/2, we can see this exception:

```
WARNING: [0x7847c7d5 0x24b549d1] Request failed: HttpPrologue[protocol=HTTP, protocolVersion=2.0, method=GET, uriPath=/static/header.html, query=, fragment=], cannot send error response, as
  response already sent
  io.helidon.http.RequestException: newPosition > limit: (83 > 40)
        at io.helidon.http.RequestException$Builder.build(RequestException.java:157)
        at io.helidon.webserver.http.ErrorHandlers.unhandledError(ErrorHandlers.java:205)
        at io.helidon.webserver.http.ErrorHandlers.lambda$handleError$1(ErrorHandlers.java:180)
        at java.base/java.util.Optional.ifPresentOrElse(Unknown Source)
        at io.helidon.webserver.http.ErrorHandlers.handleError(ErrorHandlers.java:179)
        at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:113)
        at io.helidon.webserver.http.Filters.filter(Filters.java:83)
        at io.helidon.webserver.http.HttpRoutingImpl.route(HttpRoutingImpl.java:74)
        at io.helidon.webserver.http2.Http2ServerStream.handle(Http2ServerStream.java:603)
        at io.helidon.webserver.http2.Http2ServerStream.run(Http2ServerStream.java:318)
        at io.helidon.webserver.http2.Http2Connection$StreamRunnable.run(Http2Connection.java:906)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPerTaskExecutor$ThreadBoundFuture.run(Unknown Source)
        at java.base/java.lang.VirtualThread.run(Unknown Source)
  Caused by: java.lang.IllegalArgumentException: newPosition > limit: (83 > 40)
        at java.base/java.nio.Buffer.createPositionException(Unknown Source)
        at java.base/java.nio.Buffer.position(Unknown Source)
        at java.base/java.nio.ByteBuffer.position(Unknown Source)
        at java.base/sun.nio.ch.IOUtil.write(Unknown Source)
        at java.base/sun.nio.ch.IOUtil.write(Unknown Source)
        at java.base/sun.nio.ch.SocketChannelImpl.implWrite(Unknown Source)
        at java.base/sun.nio.ch.SocketChannelImpl.write(Unknown Source)
        at io.helidon.common.socket.NioSocket.write(NioSocket.java:188)
        at io.helidon.common.socket.TlsNioSocket.wrapAndSend(TlsNioSocket.java:430)
        at io.helidon.common.socket.TlsNioSocket.doWrite(TlsNioSocket.java:389)
        at io.helidon.common.socket.TlsNioSocket.write(TlsNioSocket.java:192)
        at io.helidon.common.socket.SocketWriter.writeNow(SocketWriter.java:81)
        at io.helidon.http.http2.Http2ConnectionWriter.noLockWrite(Http2ConnectionWriter.java:200)
        at io.helidon.http.http2.Http2ConnectionWriter.writeHeaders(Http2ConnectionWriter.java:93)
        at io.helidon.http.http2.Http2ConnectionWriter.writeHeaders(Http2ConnectionWriter.java:147)
        at io.helidon.webserver.http2.Http2ServerStream.writeHeadersWithData(Http2ServerStream.java:424)
        at io.helidon.webserver.http2.Http2ServerResponse.send(Http2ServerResponse.java:156)
        at io.helidon.webserver.http2.Http2ServerResponse.send(Http2ServerResponse.java:99)
        at com.httparena.StaticHandler.handle(StaticHandler.java:72)
        at io.helidon.webserver.http.HttpRoutingImpl$RoutingExecutor.doRoute(HttpRoutingImpl.java:174)
        at io.helidon.webserver.http.HttpRoutingImpl$RoutingExecutor.call(HttpRoutingImpl.java:132)
        at io.helidon.webserver.http.HttpRoutingImpl$RoutingExecutor.call(HttpRoutingImpl.java:110)
        at io.helidon.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:76)
        ... 9 more
```

--- a/common/socket/src/main/java/io/helidon/common/socket/TlsNioSocket.java
+++ b/common/socket/src/main/java/io/helidon/common/socket/TlsNioSocket.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2025 Oracle and/or its affiliates.
+ * Copyright (c) 2025, 2026 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -188,10 +188,14 @@ public final class TlsNioSocket extends NioSocket {
 
     @Override
     public void write(BufferData buffer) {
+        // Handshake/closure and normal writes reuse the same TLS staging buffers.
+        handshakeLock.lock();
         try {
             doWrite(buffer);
         } catch (IOException e) {
             throw new UncheckedIOException(e);
+        } finally {
+            handshakeLock.unlock();
         }
     }
 

--- a/common/socket/src/test/java/io/helidon/common/socket/TlsNioSocketTest.java
+++ b/common/socket/src/test/java/io/helidon/common/socket/TlsNioSocketTest.java
@@ -1,0 +1,263 @@
+/*
+ * Copyright (c) 2026 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.socket;
+
+import java.io.IOException;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketOption;
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+import java.nio.channels.spi.SelectorProvider;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLSession;
+
+import io.helidon.common.buffers.BufferData;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class TlsNioSocketTest {
+
+    @Test
+    void closeDoesNotRaceWithInFlightTlsWrite() throws Exception {
+        BlockingSocketChannel channel = new BlockingSocketChannel();
+        SSLEngine engine = mock(SSLEngine.class);
+        SSLSession session = mock(SSLSession.class);
+        AtomicBoolean closing = new AtomicBoolean(false);
+        AtomicBoolean closeOverflowTriggered = new AtomicBoolean(false);
+        CountDownLatch closeWrapStarted = new CountDownLatch(1);
+
+        when(engine.getSession()).thenReturn(session);
+        when(session.getPacketBufferSize()).thenReturn(1);
+        when(session.getApplicationBufferSize()).thenReturn(1);
+        when(engine.wrap(any(ByteBuffer.class), any(ByteBuffer.class))).thenAnswer(invocation -> {
+            ByteBuffer src = invocation.getArgument(0);
+            ByteBuffer dst = invocation.getArgument(1);
+
+            if (!closing.get()) {
+                int consumed = Math.min(1, src.remaining());
+                for (int i = 0; i < consumed; i++) {
+                    src.get();
+                    dst.put((byte) 0x5A);
+                }
+                return new SSLEngineResult(SSLEngineResult.Status.OK,
+                                           SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING,
+                                           consumed,
+                                           consumed);
+            }
+
+            closeWrapStarted.countDown();
+            if (closeOverflowTriggered.compareAndSet(false, true)) {
+                return new SSLEngineResult(SSLEngineResult.Status.BUFFER_OVERFLOW,
+                                           SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING,
+                                           0,
+                                           0);
+            }
+
+            return new SSLEngineResult(SSLEngineResult.Status.CLOSED,
+                                       SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING,
+                                       0,
+                                       0);
+        });
+        doAnswer(invocation -> {
+            closing.set(true);
+            return null;
+        }).when(engine).closeOutbound();
+
+        TlsNioSocket socket = TlsNioSocket.server(channel, engine, "listener", "server");
+        AtomicReference<Throwable> writeFailure = new AtomicReference<>();
+        AtomicReference<Throwable> closeFailure = new AtomicReference<>();
+
+        Thread writeThread = new Thread(() -> {
+            try {
+                socket.write(BufferData.create(new byte[] {1}));
+            } catch (Throwable t) {
+                writeFailure.set(t);
+            }
+        }, "tls-write-thread");
+
+        Thread closeThread = new Thread(() -> {
+            try {
+                socket.close();
+            } catch (Throwable t) {
+                closeFailure.set(t);
+            }
+        }, "tls-close-thread");
+
+        writeThread.start();
+        assertTrue(channel.awaitWriteStarted(), "Timed out waiting for the TLS write to reach SocketChannel.write");
+
+        closeThread.start();
+        assertFalse(closeWrapStarted.await(200, TimeUnit.MILLISECONDS),
+                    "TLS close should not reach wrapAndSend while a write is still in flight");
+
+        channel.allowWriteToFinish();
+
+        writeThread.join(5_000);
+        closeThread.join(5_000);
+
+        assertTrue(closeWrapStarted.await(5, TimeUnit.SECONDS), "Close should proceed once the write finishes");
+        assertNull(writeFailure.get(), "TLS write must not fail when close races with it");
+        assertNull(closeFailure.get(), "TLS close must not fail when a write is in flight");
+    }
+
+    private static final class BlockingSocketChannel extends SocketChannel {
+        private final CountDownLatch writeStarted = new CountDownLatch(1);
+        private final CountDownLatch allowWriteToFinish = new CountDownLatch(1);
+
+        private BlockingSocketChannel() {
+            super(SelectorProvider.provider());
+        }
+
+        boolean awaitWriteStarted() throws InterruptedException {
+            return writeStarted.await(5, TimeUnit.SECONDS);
+        }
+
+        void allowWriteToFinish() {
+            allowWriteToFinish.countDown();
+        }
+
+        @Override
+        public int write(ByteBuffer src) {
+            int position = src.position();
+            int bytesToWrite = src.remaining();
+            writeStarted.countDown();
+            try {
+                if (!allowWriteToFinish.await(5, TimeUnit.SECONDS)) {
+                    throw new IllegalStateException("Timed out waiting to release blocked write");
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new IllegalStateException("Interrupted while waiting to finish write", e);
+            }
+            src.position(position + bytesToWrite);
+            return bytesToWrite;
+        }
+
+        @Override
+        public long write(ByteBuffer[] srcs, int offset, int length) {
+            long totalWritten = 0;
+            for (int i = offset; i < offset + length; i++) {
+                if (!srcs[i].hasRemaining()) {
+                    continue;
+                }
+                totalWritten += write(srcs[i]);
+                break;
+            }
+            return totalWritten;
+        }
+
+        @Override
+        public int read(ByteBuffer dst) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public long read(ByteBuffer[] dsts, int offset, int length) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public SocketChannel bind(SocketAddress local) {
+            return this;
+        }
+
+        @Override
+        public <T> SocketChannel setOption(SocketOption<T> name, T value) {
+            return this;
+        }
+
+        @Override
+        public <T> T getOption(SocketOption<T> name) {
+            return null;
+        }
+
+        @Override
+        public Set<SocketOption<?>> supportedOptions() {
+            return Collections.emptySet();
+        }
+
+        @Override
+        public SocketChannel shutdownInput() {
+            return this;
+        }
+
+        @Override
+        public SocketChannel shutdownOutput() {
+            return this;
+        }
+
+        @Override
+        public Socket socket() {
+            return null;
+        }
+
+        @Override
+        public boolean isConnected() {
+            return true;
+        }
+
+        @Override
+        public boolean isConnectionPending() {
+            return false;
+        }
+
+        @Override
+        public boolean connect(SocketAddress remote) {
+            return true;
+        }
+
+        @Override
+        public boolean finishConnect() {
+            return true;
+        }
+
+        @Override
+        public SocketAddress getRemoteAddress() {
+            return null;
+        }
+
+        @Override
+        public SocketAddress getLocalAddress() {
+            return null;
+        }
+
+        @Override
+        protected void implCloseSelectableChannel() throws IOException {
+        }
+
+        @Override
+        protected void implConfigureBlocking(boolean block) throws IOException {
+        }
+    }
+}

